### ISSUE TITLE
How should procedure types represent anonymous formals?

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#20579]
In #20554, I've assumed that when a user constructs a procedure type,
unless a formal is explicitly named (e.g., `x: int`), the formal
is anonymous. This differs from normal procedure declaration in a
big way. Consider:

```chapel
// Here the formals are 'x' and 'y', and they are of type 'int'.
proc foo(x: int, y: int): int { return x + y; }

// Here the formals are 'x' and 'y', and they are of type 'any'.
proc foo(x, y): int { return x + y; }
```

The key thing to note is that for any procedure definition, when
a formal is specified without a type (e.g., a single identifier `x`),
the symbol refers to the formal _name_, and not its type. Its type
is implicitly `any`, and the formal is generic. (The `any` type is
modelled in the compiler(s) today, but I do not think it is present
in the language spec.)

Contrast that to procedure type construction in my PR:

```chapel
// Here the formals are named, and they are both of type 'int'.
type F1 = proc(x: int, y: int): int;

// Here the formals are anonymous, and they are of type 'int'.
type F2 = proc(int, int): int;
```

Right now, a formal in a procedure type that is a single identifier
is treated as the _type_ of an anonymous formal.

It is also impossible to specify that a formal is generic and has
the type `any`.

This incongruence could be seen as a problem because the meaning
of an otherwise unannotated formal differs.

## Proposal 1: represent an anonymous formal with '_'

In this proposal we align procedure definition and procedure types
by representing an anonymous formal with `_`. Now it becomes
possible to express procedure types in the exact same way as
procedure definitions. The only learning curve is that users can
express an anonymous formal using `_`.

```chapel
// Here the formals are named, and they are both of type 'int'.
type F1 = proc(x: int, y: int): int;

// Here the formals are anonymous, and they are of type 'int'.
type F2 = proc(_: int, _: int): int;

// Here the formals are named, and they are of type 'any'.
type F3 = proc(x, y): int;

// Here the formals are anonymous, and both of type 'any'.
type F4 = proc(_, _): int;
```

This also avoids any need to discuss introducing the `any` type.

## Proposal 2: document the differences and introduce the `any` type

In this proposal we clearly document that unless fully annotated
(e.g., `x: int`, formals in procedure types are anonymous. This
gains us the advantage of being able to represent the (arguably)
most common use case for procedure types in the must succinct way.

Unless we introduce the `any` type or some syntactic equivalent,
it will not be possible for users to construct a procedure type
where a formal is generic.

Thus we can introduce `any`. Here I've opted to represent it with
the `?` symbol to avoid introducing a new reserved word:

```chapel
// Here the formals are named, and they are both of type 'int'.
type F1 = proc(x: int, y: int): int;

// Here the formals are anonymous, and they are of type 'int'.
type F2 = proc(int, int): int;

// Here the formals are named, and they are of type 'any'.
type F3 = proc(x: ?, y: ?): int;

// Here the formals are anonymous, and both are of type 'any'.
type F4 = proc(?, ?): int;
```

**Comment by benharsh** [comment]
What if we used ``?`` to represent an anonymous formal's "name":
```chpl
type FN = proc(?:int, ?:int) : int;
```
A downside of this approach would be that ``?`` is primarily used with type expressions today (perhaps exclusively?), and it might confuse users to see it in other contexts.

-----

I also think that a decision here should be consistent with our choices for similar situations involving loop iterands and multi-decls:
```chpl
var (x, y, _) = foo(...); // returns 3*int

// Use a leader with a sophisticated pattern, whose yielded values we do not care about
forall (_, x, y) in zip(Conductor, LHS, RHS) do ...;
```

**Comment by dlongnecke-cray** [comment]
Yeah, I think we should probably use `_` over `?` for an anonymous formal's name unless we're contemplating changing tuple unpacking as well (which AFAIK we're pretty happy with). I think `_` is perfect because as you say, `?` is used in a type context today, and secondly, users are already used to `_` as meaning "the name of this particular component doesn't matter".

In my proposal 2 I advocated using `?` to represent any type. I'm now kind of on the fence about it, but maybe it's ok because even if it looks like a type query, it doesn't have an identifier hugging it, and secondly I think we might be having conversations about letting `?` represent `any` in other contexts.

**Comment by jeremiah-corrado** [comment]
I like proposal 1 better.  

In addition to the points that have already been made, I think the meaning of:
```chapel
type F2 = proc(_: int, _: int): int;
```
is much more visually obvious than:
```chapel
type F2 = proc(int, int): int;
```

Specifically, I like that the first case is explicitly saying: "I want two ints, but I don't care what their names are", rather than: "I want two ints"

**Comment by bradcray** [comment]
I agree with what's been said so far and don't think `proc (x, x)` should be interpreted as "two arguments of type x".  I'm very comfortable with proposal 1.  I'd also be comfortable with `proc (:x, :x)` to make the `x`'s more obviously a type constraint, but I think asking people to type `_` as a placeholder for the argument name is not too much to ask, and makes the syntactic pattern clearer and more obvious.

I wanted to check on this comment from @benharsh :

> I also think that a decision here should be consistent with our choices for similar situations involving loop iterands and multi-decls:

To see whether, Ben, you're saying that you think proposal 1 is arguably consistent with existing uses of `_` or is not in some way?  (at a high-level it seems consistent in a "don't care to name it" sort of way, but maybe you were seeing something more subtle).

I'm also comfortable with using `?` as an "any type" indicator in this context, but am not entirely clear how that would work in practice.  In most contexts that need FCFs, don't we need to know the types in order to be able to make the call?  Or is the idea that this would result in some sort of templating capability where we'd stamp out different versions of the routine for each procedure signature that was passed in?

**Comment by benharsh** [comment]
> To see whether, Ben, you're saying that you think proposal 1 is arguably consistent with existing uses of _ or is not in some way? (at a high-level it seems consistent in a "don't care to name it" sort of way, but maybe you were seeing something more subtle).

That's all I was saying: that we shouldn't have different ways of indicating "I don't care to name it".

I also lean towards proposal 1, which feels more consistent with procs in general. Especially in the case where I want to create an anonymous function to pass into some callback handler:

```chpl
var mc = new MyCollection(string);
// ... fill 'mc' from a file ...

MyCollection.update(proc(idx:int, val : string) {
  val = val.strip();
});
```

It's not quite as nice when creating a type-expression for a first-class function, but it's something we can live with. I'd like to see if we can make something like ``proc(:int,:int)`` work.

**Comment by dlongnecke-cray** [comment]
TLDR: I think that an unannotated formal in a function type should
represent the formal's type and not its name.

Anecdotal stuff first - when I set out to add this feature, I always
intended to have an unannotated formal represent a type, and until
@mstrout pointed out the inconsistency with named functions to me it
was actually inconceivable to me that an unannotated formal in a type
context could mean anything else.

This is because a lot of other languages do the same thing in a type
context. The current "formal is type" interpretation is in many ways
"how I was raised". I know that's not exactly the most compelling of
arguments, and if we always followed that sort of reasoning we'd
never see improvement, but...

The Rust grammar allows users to specify formal names in function
types.

```rust
type T = fn(x: i64, y: i64) -> i64;
```

The syntax for fully annotated formals actually looks very similar to
what it looks like in Chapel:

```chapel
type T = proc(x: int, y: int): int;
```

However in Rust, an unannotated formal represents a formal's type,
and not its name:

```rust
type T = fn(i32, i32) -> i32;

fn call(f: T) -> i32 { return f(4, 4); }

fn add(x: i32, y: i32) -> i32 { return x + y; }

fn main() {
  let x = call(add);
  assert!(x == 8);
}
```

Unannotated formals in a function type in Swift also represent the
formal's type:

```swift
typealias T = (Int, Int) -> Int

func call(_ f: T) -> Int { return f(4, 4) }

func add(x: Int, y: Int) -> Int { return x + y }

func main() {
  let x = call(add)
  assert(x == 8)
}
```

Here's what the Swift compiler tells me if I try to name formals in
a function type:

```swift
typealias T = (x: Int) -> Int
```

```
TestSwift.swift:3:16: error: function types cannot have argument
labels; use '_' before 'x'
typealias T = (x: Int) -> Int
```

Note that using a `_` in an argument label in Swift effectively
makes a formal anonymous - its name can no longer be used in named
argument passing. So the Swift compiler is going so far as to tell
me explicitly that formals in function types cannot have names that
are meaningful w.r.t. named argument passing (though you can _have_
named formals).

It is strongly engrained in my brain to reason about formal types in a
type context. I might have gotten this from C:

```c
typedef int(*T)(int, int);
```

Where formal names in function types are optional labels that can be
discarded.

Swift and Rust might have tried the alternative and found it to be too
complicated/circumstantial to be of value - or maybe like
with Chapel and switching to [0] based indexing, they opted not to go
against the tide. Or maybe they never even considered it, IDK yet.

---

Additionally, there is the matter of convenience. I would like to
think that in most scenarios, users will want to reason about
the formals of a function type without regard for its names.
I think it would be very irritating if I had to type something like:
`proc(_: int, _: int): int` every time I wanted to ask for a function
that takes two `int` and returns an `int`.

I would wager this to be a reasonable default in the same way
that named argument passing is _not_ the default (but it _is_
in Swift! How horrifying, could you imagine...). If most calls are
made without named arguments, why are we prioritizing the
clarity of named arguments (especially if we don't even know
yet if they'll participate in function types)?

Shortening it to `proc(:int, :int)` goes against my personal tastes,
as it's starting to look a bit too cramped inside that formal
list and could use a bit of room to breathe (we're not PHP!). As
well, the `:` is not very legible to me.

I think convenience is secondary to confusion, though. It would take
me a long while to get used to the alternative meaning, and
I expect I'd keep trying to type `proc(int, int): int`.

**Comment by bradcray** [comment]
Despite the precedent in Rust and Swift, I'm still disinclined to have an
undecorated `t` in a `proc`'s argument list indicate a type constraint
rather than an argument name.  The precedent in C-like languages seems
easier to ignore since, syntactically, the type comes first in C and doesn't
get separated by a colon.  By contrast, in Chapel, this would be the
first case I can think of where I type constraint / specifier did not
follow a colon.

Making that statement more concrete with an example, the following
asymmetry concerns me:

```chapel
type x = ...;

proc foo(x) { ... }  // x is an argument name
var f: proc(x);   // x is a type name
var f2 = proc(x) { writeln(x); };  // x is an argument name
type ft = proc(x);  // x is a type name
foo(proc(x));  // x is a type name
foo(proc(x) { writeln(x); }  // x is an argument name

proc foo(type t) { writeln("In type foo"); }
proc foo(x) { writeln("In value foo"); x(3); }
```

Given that type and value expressions are so syntactically mixable in
Chapel, it feels worrisome to require the reader to think about whether
the `proc` keyword is introducing a type or value in order to interpret
the meaning of its argument list.

I also think it's easier to go from a stricter rule (you must use a colon) to
a looser one (OK, you can optionally drop it in these cases) than vice-versa,
so think we should start strict, live with it, and see how things go rather
than starting with the subtler rule.

**Comment by dlongnecke-cray** [comment]
> Chapel, it feels worrisome to require the reader to think about
whether the proc keyword is introducing a type or value in order
to interpret the meaning of its argument list.

I would argue that this ambiguity already exists.

If I call `foo(something)`, I don't know without further context if
`something` is a type or a value.

In the `foo(proc(x))` case, I know that `proc(x)` is a type, because
syntactically function types are not followed by a body. Likewise, I
know that `proc(x) { writeln(x); }` is a value, because it has a body.
So there is enough syntactic info available to draw the dinstiction.

It's frustrating to me because it seems like the main reason we are
hesitant to adopt this style is because `any` type formals exist.
If we lived in a world where you had to write `proc foo(x: ?)` to
indicate that the formal `x` was generic, would the "formal is type
in type context" proposal seem more reasonable?

Rust has a similar problem with its closure syntax:

```rust
type T = fn(i32, i32) -> i32;
let f = |x, y| x + y;
println!("{}", f(4, 4));
```

Here the types of `x` and `y` are elided instead of generic, but
syntactically the problem seems the same.

Perhaps the reason why this is acceptable is owing to the presence
of `|` instead of `fn(...)`? The different opening syntax for the
form.

However I think expecting users to interpret the meaning of a name
differently in a type vs value context is not too big a sell - I think
that users already make such a mental distinction (at least, I do)
when reasoning about types versus values today.

> I also think it's easier to go from a stricter rule (you must use a
colon) to a looser one (OK, you can optionally drop it in these cases)
than vice-versa, so think we should start strict, live with it, and
see how things go rather than starting with the subtler rule.

It seems like we can't really do that, at least not without someone
successfully champion'ing the `x: ?` formal idea.

Because it's already a context-sensitive issue - `proc(x)` has a well
defined meaning in a value context. If we require `proc(x: int)` in a
type context, then we're already forcibly requiring the user to draw
a distinction.

**Comment by dlongnecke-cray** [comment]
All I was trying to suggest is that if we require `proc(x: int)` in a
type context, that means that on some level, both the compiler and
the user still have to be aware of the difference and can identify
whether a `proc` is a type or a value. Because I can still write
`proc(x) { ... }`, but only in a value context.

The relationship to `x: ?` idea is about _requiring_ that a generic
formal annotate its type with `:?`. I don't know if such an idea
would ever gain traction, but I'm arguing that I don't think we'd ever
consider "unannotated formal means type" unless that was the case
(because the whole basis of the `proc(_: x)` stance seems to be about
ensuring syntactic uniformity).

Basically, that there's no common ground to stand on in saying that
"we could start strict and relax later", because I don't think we
ever can - `proc(x) { ... }` has a meaning in a value context that
is unlikely to change.

**Comment by bradcray** [comment]
> The relationship to x: ? idea is about requiring that a generic formal annotate its type with :?

Oh, I don't think we should require it.  But I think it makes sense to support it as an option.

> Because I can still write `proc(x) { ... }`, but only in a value context.

Only if/when we support some form of type elision, though right?  Until then, FCFs can't be generic, so have to have their types fully specified anyway?

Even if/once we have type elision, though, I agree that starting strict means that you can only write it in a value context, where the argument would be that in all `proc [ident](x)` contexts, 'x' is an argument name.

> (because the whole basis of the proc(_: x) stance seems to be about ensuring syntactic uniformity).

I agree that that's the basis for why I'm advocating for starting strict.  But if, after using the new FCF features for some time, we and/or users revolt and say "I'm tired of typing `:` (or `_:`) before my type signifiers in all of my FCF types, the clarity isn't worth it, let us stop!" then I don't see that there's any problem in doing so.  Yes, the syntax becomes less uniform (i.e., "Brad chose incorrectly"), but it shouldn't break any working code.  Whereas if we start out supporting unannotated formals in type contexts as meaning a type constraint, and later hear from users that it's too confusing s.t. we decide to require the colon, then code that had worked before would stop working.

**Comment by dlongnecke-cray** [comment]
I've opened https://github.com/chapel-lang/chapel/pull/20788 to implement the "start strict" approach as proposed by Brad.
The long form of the error message mentions that the meaning is currently ambiguous. I wanted to go so far as to point to this issue in the compiler error, but didn't because I was certain if that was frowned upon or not.

**Comment by bradcray** [comment]
I would not have a problem about pointing to this issue in the error message, at least for a time.  That said, this issue has sort of evolved from its OP I think (?), so a fresh issue that showed the two proposals from your slides afresh might make more sense.

**Comment by dlongnecke-cray** [comment]
After conversations with users and the rest of the team, I think I can say that we will represent anonymous formals with a `_`, such that:

```chapel
type T = proc(_: int, _: int): int; // Takes two 'int', returns 'int', formal names don't matter
```

And secondly that, for the time being, all formals in a function type must be named (where `_` is a unique name taken to mean "the name of this formal does not matter").

I'm going to close this issue, but feel free to open if you disagree with the current direction.
